File: 03.programmers/lv02/01.C++/English_chain_game.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;

vector<int> solution(int n, vector<string> words) {
    vector<int> answer;
    unordered_set<string> us;
    vector<int> cnt(n,0);
    for(int i=0; i<words.size(); i++){ 
        int idx = i%n;
        string key = words[i];
        if(i>0 && words[i-1].back() != words[i].front()){
            answer.push_back(idx+1);
            answer.push_back(cnt[idx]+1);
            return answer;
        }
        if(!us.count(key)){
            us.insert(key);
            cnt[idx]++;
        }
        else{
            answer.push_back(idx+1);
            answer.push_back(cnt[idx]+1);
            return answer;
        }
    }

    return answer = {0,0};
}
int main(){
    vector<string> w = {"tank", "kick", "know", "wheel", "land", "dream", "mother", "robot","tank"};
    for(auto ans : solution(3,w))
        cout<<ans<<" ";
    return 0;
}

File: 03.programmers/lv02/02.Python/English_chain_game.py
def solution(n, words):
    for i in range(1, len(words)):
        if words[i-1][-1] != words[i][0] or words[i] in words[:i]: return [(i%n)+1, (i//n)+1]
    else: return [0,0]

input_n = 3
input_words = ["tank", "kick", "know", "wheel", "land",
                "dream", "mother", "robot", "tank"]
# input_n = 2
# input_words = ["hello", "one", "even", "never", "now", "world", "draw"]

print(solution(input_n, input_words))
